pkgs/native_assets_cli/lib/src/model/link_config.dart

```diff
@@ -0,0 +1,165 @@
+class LinkConfigImpl extends HookConfigImpl implements LinkConfig {
+  @override
+  Hook get hook => Hook.link;
+
+  @override
+  final List<AssetImpl> assets;
+
+  final BuildConfigImpl _buildConfig;
+
+  @override
+  final List<Resource>? resources;
+
+  final _LinkConfigArgs _args;
+
+  LinkConfigImpl._(
+    this._args, {
+    required this.assets,
+    required BuildConfigImpl buildConfig,
+    required ResourceIdentifiers? resourceIdentifiers,
+  })  : _buildConfig = buildConfig,
+        resources = fromIdentifiers(resourceIdentifiers);
+
+  @override
+  Uri get outputDirectory => _buildConfig.outputDirectory;
+
+  @override
+  String get outputName => 'link_output.json';
+
+  @override
+  String get packageName => _buildConfig.packageName;
+
+  @override
+  Uri get packageRoot => _buildConfig.packageRoot;
+
+  @override
+  String toJsonString() =>
+      const JsonEncoder.withIndent('  ').convert(_args.toJson());
+
+  @override
+  Version get version => _buildConfig.version;
+
+  static LinkConfig fromArguments(List<String> arguments) =>
+      _LinkConfigArgs.fromArguments(arguments).toLinkConfig();
+
+  factory LinkConfigImpl.fromValues({
+    required Uri? resourceIdentifierUri,
+    required BuildConfigImpl buildConfig,
+    required List<AssetImpl> assetsForLinking,
+  }) =>
+      _LinkConfigArgs(
+        assetsForLinking: assetsForLinking,
+        buildConfig: buildConfig,
+        resourceIdentifierUri: resourceIdentifierUri,
+      ).toLinkConfig();
+
+  @override
+  BuildMode get buildMode => _buildConfig.buildMode;
+
+  @override
+  CCompilerConfig get cCompiler => _buildConfig.cCompiler;
+
+  @override
+  bool get dryRun => _buildConfig.dryRun;
+
+  @override
+  Iterable<String> get supportedAssetTypes => _buildConfig.supportedAssetTypes;
+
+  @override
+  int? get targetAndroidNdkApi => _buildConfig.targetAndroidNdkApi;
+
+  @override
+  Architecture? get targetArchitecture => _buildConfig.targetArchitecture;
+
+  @override
+  IOSSdk get targetIOSSdk => _buildConfig.targetIOSSdk;
+
+  @override
+  OS get targetOS => _buildConfig.targetOS;
+
+  /// The version of [BuildConfigImpl].
+  ///
+  /// This class is used in the protocol between the Dart and Flutter SDKs
+  /// and packages through build hook invocations.
+  ///
+  /// If we ever were to make breaking changes, it would be useful to give
+  /// proper error messages rather than just fail to parse the JSON
+  /// representation in the protocol.
+  static Version latestVersion = Version(1, 0, 0);
+}
+
+List<Resource>? fromIdentifiers(ResourceIdentifiers? resourceIdentifiers) =>
+    resourceIdentifiers?.identifiers
+        .map((e) => Resource(name: e.name, metadata: e.id))
+        .toList();
+
+class _LinkConfigArgs {
+  final Uri? resourceIdentifierUri;
+  final BuildConfigImpl buildConfig;
+  final List<AssetImpl> assetsForLinking;
+
+  static const resourceIdentifierKey = 'resource_identifiers';
+  static const buildConfigKey = 'build_config';
```

Comment:
Why do we have a `build_config` key? I would expect to simply have a subset of the build config toplevel keys (and then in addition the `resource_identifiers` and `assets` toplevel keys).
Why is there a need for assymmetry?
Since our Dart API is unified with HookConfig, the serialization should also be unified.

Reply:
I would not want to duplicate the complex serialization from `BuildConfig`. But I am fine with a rewrite in a future PR :)

Reply:
Can't all the shared things be moved into HookConfig?

Reply:
It would also mean that things that are not visible in the LinkConfig, are still in the JSON file and if these change the caching doesn't work?

Reply:
Probably they can. I think this would require quite some changes though, which can be done decoupled from the other changes in this PR, so I would defer that to a future PR.

Reply:
That doesn't work. We can't change the layout of `link_config.json` in a breaking way, because we don't have protocol negotiation. The Dart/Flutter SDK does not know what version of the protocol the link hook invoked speak.